enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,445 @@
+
+When the value of `IngressPlacement` is `Workers`, the defaulting behavior of
+`.spec.replicas` and `.spec.nodePlacement` will be the same as it is today:
+`.spec.replicas` will be chosen according to the value of
+`InfrastructureTopology`, namely `1` when `SingleReplica` or `2` when
+`HighlyAvailable`. `.spec.nodePlacement` will always just be:
+
+```yaml
+nodePlacement:
+  nodeSelector:
+    matchLabels:
+      kubernetes.io/os: linux
+      node-role.kubernetes.io/worker: ''
+```
+
+However, if the value of `IngressPlacement` is `ControlPlane`, the defaulting
+behavior will be different: `.spec.replicas` will be chosen instead according
+to the value of `ControlPlaneTopology`; again, `1` when `SingleReplica` or `2`
+when `HighlyAvailable`. `.spec.nodePlacement` will be always just be:
+
+```yaml
+nodePlacement:
+  nodeSelector:
+    matchLabels:
+      kubernetes.io/os: linux
+      node-role.kubernetes.io/master: ''
+```
+
+(Note that the `kubernetes.io/os: linux` label is mentioned just because it's
+the current behavior, it has no importance in this enhancement)
+
+The installer will detect situations in which it's unlikely the user will want
+to set up a load-balancer. For now, those situations only include installation
+of single control-plane node cluster deployments on the "none" platform. In
+those situations, the installer will set `IngressPlacement` to be
+`ControlPlane`. Since there's just a single control-plane node, `ControlPlane`
+topology would be `SingleReplica` and the combined effect would be that the
+`IngressController` will have just a single replica and be pinned to the single
+control-plane node. This will then ensure that the `router-default` deployment
+created by the Cluster Ingress Operator will always run on the single
+control-plane node, and as a result any `*.apps.<cluster>.<base>` DNS entries
+which originally pointed at the single control-plane node will remain correct
+even in the face of newly added worker nodes.
+
+In any other situations, the installer will set `IngressPlacement` to
+`Workers`, resulting in the same default behavior as before this enhancement,
+namely that `IngressController` pod replicas are scheduled on worker nodes and
+determined according to the `InfrastructureTopology`.
+
+In the future, when IPI-installed single control-plane node clusters in the
+cloud no longer provision a load-balancer by default, they would also benefit
+from having the installer set the `IngressPlacement` to `ControlPlane`.
+
+### Risks and Mitigations
+
+This should make no noticable difference on "regular" single control-plane node
+clusters which do not have any day-1 or day-2 worker nodes. The only difference
+for those clusters would be the `IngressController` targeting the "master" pool
+rather than the "worker" pool, but since the single control-plane node is
+already both in the "master" and "worker" pools, that should make no practical
+difference.
```

Comment:
Is there any impact on multi-node clusters? The default behavior won't change, but is there any reason to be concerned if the user changes the setting to "ControlPlane" themselves? Can they break the cluster?

Reply:
I think for now we're not gonna let the user set this parameter, so there's no risk of it being applied in non-single-node clusters.

Having said that, hypothetically if the user was able to set this new parameter to ControlPlane on non-single-node clusters, especially ones that have installer-provisioned load-balancer. I think it does have the *potential* to mess things up - it would depend heavily on how the load balancer is being set up to find out which nodes it should target.

I don't know the details in reality, so maybe it's not an actual problem, but I imagine this hypothetical scenario -

- User chooses 3 control-plane 3 worker cluster installation on <some potentially problematic platform>
- User chooses IngressPlacement defaulting behavior to target ControlPlane
- Platform's installer naively points the provisioned load balancer only at the worker nodes (since it sees that this is not a compact cluster where control-plane nodes are also considered workers), that platform's installer does that because it's completely unaware of the effects of the IngressPlacement parameter* 
- IngressController gets created with 2 replicas (because master topology is HA) and it would node-select the master pool, 2 router-default pod gets placed on 2 random control-plane nodes
- Ingress traffic is broken because the load balancer is not pointing at the correct nodes

The bullet marked with * at the end of it is where I have a bit of uncertainty - are any platform installer's doing this "naive" action of pointing the load balancer only at the cluster's worker nodes? If yes, this is gonna be a problem that's gonna have to be solved whenever we make this parameter user-configurable

Reply:
> I think for now we're not gonna let the user set this parameter, so there's no risk of it being applied in non-single-node clusters.

How will you enforce this?

Reply:
>How will you enforce this?

See https://github.com/openshift/enhancements/pull/1041#discussion_r820636797

Reply:
> The installer does not create the load balancer that ingress uses.

It does for at least one platform, namely Azure (see [BZ#1794839](https://bugzilla.redhat.com/show_bug.cgi?id=1794839)).  However, in the case of Azure, the load balancer always includes both control-plane as well as worker nodes.

Reply:
Yes, you are technically correct that, in the case of Azure, the installer creates the load balancer used by ingress. However, the installer does not create the backends, which is the part of the load balancer that is germane to this discussion.